### ISSUE TITLE
no-redeclare should support ES6 block scope

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#2337]
```js
// test.js
switch (foo) {
    case true:
        let bar = 42;
        break;
    case false:
        let bar = 255;
        break;
    default:
        const bar = foo;
}
```

```js
{
    "ecmaFeatures": {
        "blockBindings": true
    },
    "rules": {
        "no-redeclare": 2
    }
}
```

```sh
$ eslint --reset --no-eslintrc -c config.json test.js

# No output
```

Since the `case` block of `switch` statements is a single lexical environment, `let` and `const` as in the above test should trigger warnings. (Running the test code in Firefox triggers a TypeError.)

**Comment by gyandeeps** [comment]
TypeError is explained here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#Temporal_dead_zone_and_errors_with_let

**Comment by btmills** [comment]
Looks like `no-redeclare` just hasn't been updated for ES6 block scoping yet. I'll rename the issue accordingly.

**Comment by nzakas** [comment]
Working on this.
